File: kong-plugin-ascend/kong/plugins/ascend/handler.lua
--[[
ASCEND Kong Plugin - Request Handler
====================================

Main plugin logic implementing AI agent governance.

FAIL SECURE DESIGN:
- All errors result in request denial (403)
- Only explicitly approved actions proceed
- Circuit breaker prevents cascade failures

Plugin Lifecycle:
- init_worker: Initialize shared state
- access: Main authorization logic (blocks request if needed)
- log: Async metrics and logging

Author: ASCEND Platform Engineering
--]]

local constants = require "kong.plugins.ascend.constants"
local http_client  -- Lazy loaded to avoid circular dependency

local kong = kong
local ngx = ngx
local cjson = require "cjson.safe"
local fmt = string.format

-- Plugin handler object
local AscendHandler = {
    PRIORITY = constants.PRIORITY,
    VERSION = constants.VERSION
}

--------------------------------------------------------------------------------
-- Utility Functions
--------------------------------------------------------------------------------

--- Generate cache key from request attributes
-- @param agent_id Agent identifier
-- @param method HTTP method
-- @param path Request path
-- @return string Cache key
local function generate_cache_key(agent_id, method, path)
    -- Normalize path by replacing numeric IDs with placeholder
    local normalized_path = ngx.re.gsub(path, "/[0-9]+(/|$)", "/{id}$1", "jo")
    -- Also handle UUIDs
    normalized_path = ngx.re.gsub(normalized_path,
        "/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}(/|$)",
        "/{id}$1", "ijo")

    return fmt("%s%s:%s:%s", constants.CACHE_PREFIX, agent_id, method, normalized_path)
end

--- Map HTTP method to action type
-- @param method HTTP method (GET, POST, etc.)
-- @return string Base action type
local function method_to_action(method)
    return constants.METHOD_ACTION_MAP[method] or "unknown"
end

--- Determine action type from path patterns
-- @param path Request path
-- @param method HTTP method
-- @param custom_mappings Custom action type mappings from config
-- @return string Action type
local function determine_action_type(path, method, custom_mappings)
    -- Check custom mappings first
    if custom_mappings then
        for pattern, action_type in pairs(custom_mappings) do
            if ngx.re.match(path, pattern, "ijo") then
                return action_type
            end
        end
    end

    -- Check built-in resource patterns
    for _, mapping in ipairs(constants.RESOURCE_PATTERNS) do
        if ngx.re.match(path, mapping.pattern, "ijo") then
            return mapping.action
        end
    end

    -- Default: method-based action with resource name
    local base_action = method_to_action(method)
    local resource = ngx.re.match(path, "^/([^/]+)", "jo")

    if resource and resource[1] then
        return fmt("%s_%s", base_action, resource[1])
    end

    return base_action
end

--- Check if path is excluded from governance
-- @param path Request path
-- @param excluded_paths List of excluded path patterns
-- @return boolean True if excluded
local function is_path_excluded(path, excluded_paths)
    if not excluded_paths or #excluded_paths == 0 then
        return false
    end

    for _, pattern in ipairs(excluded_paths) do
        if ngx.re.match(path, pattern, "ijo") then
            kong.log.debug("Path excluded from governance: ", path)
            return true
        end
    end

    return false
end

--- Extract agent ID from request
-- @param conf Plugin configuration
-- @return string|nil Agent ID or nil if not found
-- @return string|nil Error message if agent ID required but missing
local function extract_agent_id(conf)
    local agent_id = kong.request.get_header(constants.HEADERS.AGENT_ID)

    if agent_id then
        return agent_id, nil
    end

    -- Use default if configured
    if conf.default_agent_id then
        return conf.default_agent_id, nil
    end

    -- Check if required
    if conf.require_agent_id then
        return nil, "Missing required header: " .. constants.HEADERS.AGENT_ID
    end

    return nil, nil
end

--- Build request context for ASCEND API
-- @param conf Plugin configuration
-- @return table Request context
local function build_request_context(conf)
    local ctx = {
        source_ip = kong.client.get_ip(),
        request_id = kong.request.get_header(constants.HEADERS.REQUEST_ID) or ngx.var.request_id,
        timestamp = ngx.now() * 1000,  -- milliseconds
        gateway = "kong",
        gateway_version = kong.version
    }

    -- Add organization if configured
    if conf.organization_id then
        ctx.organization_id = conf.organization_id
    end

    return ctx
end

--- Build ASCEND API request payload
-- @param agent_id Agent identifier
-- @param action_type Action type
-- @param conf Plugin configuration
-- @return table API request payload
local function build_api_payload(agent_id, action_type, conf)
    local payload = {
        agent_id = agent_id,
        action_type = action_type,
        environment = conf.environment,
        data_sensitivity = conf.data_sensitivity,
        context = build_request_context(conf)
    }

    -- Optionally include request details
    if conf.include_request_body then
        local body, err = kong.request.get_body()
        if body and not err then
            -- Truncate if too large
            local body_str = cjson.encode(body)
            if body_str and #body_str <= conf.max_body_size then
                payload.request_body = body
            end
        end
    end

    return payload
end

--- Send deny response
-- @param status_code HTTP status code
-- @param error_code Error code constant
-- @param message Human-readable message
-- @param extra Additional response fields
local function deny_request(status_code, error_code, message, extra)
    local response = {
        error = error_code,
        message = message,
        timestamp = ngx.now() * 1000
    }

    -- Merge extra fields
    if extra then
        for k, v in pairs(extra) do
            response[k] = v
        end
    end

    kong.log.info(fmt("ASCEND DENY: %s - %s", error_code, message))

    return kong.response.exit(status_code, response, {
        ["Content-Type"] = "application/json",
        [constants.HEADERS.DECISION] = "denied"
    })
end

--- Allow request to proceed with governance headers
-- @param decision ASCEND decision response
local function allow_request(decision)
    -- Add governance headers for downstream services
    kong.service.request.set_header(constants.HEADERS.DECISION, "approved")
    kong.service.request.set_header(constants.HEADERS.ACTION_ID, tostring(decision.id or ""))

    if decision.risk_score then
        kong.service.request.set_header(constants.HEADERS.RISK_SCORE, tostring(decision.risk_score))
    end

    if decision.risk_level then
        kong.service.request.set_header(constants.HEADERS.RISK_LEVEL, decision.risk_level)
    end

    kong.log.info(fmt("ASCEND ALLOW: action_id=%s risk_score=%.1f",
        decision.id or "unknown", decision.risk_score or 0))
end

--------------------------------------------------------------------------------
-- Plugin Lifecycle Handlers
--------------------------------------------------------------------------------

--- Initialize worker process
function AscendHandler:init_worker()
    kong.log.info("ASCEND plugin initialized - FAIL SECURE mode")

    -- Lazy load HTTP client
    http_client = require "kong.plugins.ascend.http_client"
end

--- Main access phase handler
-- @param conf Plugin configuration
function AscendHandler:access(conf)
    local start_time = ngx.now()
    local method = kong.request.get_method()
    local path = kong.request.get_path()

    -- Store for log phase
    kong.ctx.plugin.start_time = start_time
    kong.ctx.plugin.method = method
    kong.ctx.plugin.path = path

    -- Check path exclusions
    if is_path_excluded(path, conf.excluded_paths) then
        kong.ctx.plugin.excluded = true
        return  -- Allow request
    end

    -- Extract agent ID
    local agent_id, err = extract_agent_id(conf)
    if err then
        return deny_request(
            conf.denied_status_code,
            constants.ERROR.MISSING_AGENT_ID,
            err
        )
    end

    if not agent_id then
        -- Not required, no default - allow but log warning
        kong.log.warn("No agent ID provided, skipping governance")
        kong.ctx.plugin.no_agent = true
        return
    end

    kong.ctx.plugin.agent_id = agent_id

    -- Determine action type
    local action_type = determine_action_type(path, method, conf.custom_action_types)
    kong.ctx.plugin.action_type = action_type

    -- Check cache for approved decision
    local cache_key = generate_cache_key(agent_id, method, path)

    if conf.cache_ttl > 0 then
        local cached = kong.cache:get(cache_key)
        if cached then
            kong.ctx.plugin.cache_hit = true
            kong.log.debug("Cache hit for: ", cache_key)
            allow_request(cached)
            return
        end
    end

    -- Submit to ASCEND API
    local payload = build_api_payload(agent_id, action_type, conf)
    local decision, api_err = http_client.submit_action(conf, payload)

    if api_err then
        kong.ctx.plugin.api_error = api_err

        -- FAIL SECURE: Deny on error unless fail_open configured
        if conf.fail_open then
            kong.log.warn("ASCEND API error, fail_open=true, allowing request: ", api_err)
            kong.ctx.plugin.fail_open_triggered = true
            return
        end

        return deny_request(
            conf.error_status_code,
            constants.ERROR.API_ERROR,
            "Governance service unavailable",
            { detail = api_err }
        )
    end

    kong.ctx.plugin.decision = decision

    -- Handle decision
    local status = decision.status

    if status == constants.STATUS.APPROVED then
        -- Cache approved decision
        if conf.cache_ttl > 0 then
            kong.cache:set(cache_key, decision, conf.cache_ttl)
        end

        allow_request(decision)
        return
    end

    if status == constants.STATUS.PENDING then
        -- Pending approval
        if conf.block_on_pending then
            return deny_request(
                conf.pending_status_code,
                constants.ERROR.PENDING,
                "Action requires approval",
                {
                    action_id = decision.id,
                    risk_score = decision.risk_score,
                    risk_level = decision.risk_level,
                    requires_approval = true
                }
            )
        else
            -- Allow but flag as pending
            kong.service.request.set_header(constants.HEADERS.DECISION, "pending")
            kong.service.request.set_header(constants.HEADERS.ACTION_ID, tostring(decision.id or ""))
            return
        end
    end

    if status == constants.STATUS.DENIED then
        return deny_request(
            conf.denied_status_code,
            constants.ERROR.DENIED,
            decision.denial_reason or "Action denied by governance policy",
            {
                action_id = decision.id,
                risk_score = decision.risk_score,
                risk_level = decision.risk_level,
                denial_reason = decision.denial_reason
            }
        )
    end

    -- Unknown status - FAIL SECURE
    kong.log.err("Unknown ASCEND status: ", status)
    return deny_request(
        conf.error_status_code,
        constants.ERROR.INTERNAL,
        "Invalid governance response"
    )
end

--- Log phase handler (async, after response sent)
-- @param conf Plugin configuration
function AscendHandler:log(conf)
    if not conf.log_decisions then
        return
    end

    local ctx = kong.ctx.plugin
    local latency_ms = (ngx.now() - (ctx.start_time or ngx.now())) * 1000

    -- Build log entry
    local log_entry = {
        plugin = constants.NAME,
        version = constants.VERSION,
        timestamp = ngx.now() * 1000,
        latency_ms = latency_ms,
        method = ctx.method,
        path = ctx.path,
        agent_id = ctx.agent_id,
        action_type = ctx.action_type,
        excluded = ctx.excluded or false,
        cache_hit = ctx.cache_hit or false,
        fail_open_triggered = ctx.fail_open_triggered or false
    }

    -- Add decision details
    if ctx.decision then
        log_entry.decision_status = ctx.decision.status
        log_entry.action_id = ctx.decision.id
        log_entry.risk_score = ctx.decision.risk_score
        log_entry.risk_level = ctx.decision.risk_level
    end

    -- Add error if present
    if ctx.api_error then
        log_entry.error = ctx.api_error
    end

    -- Log as JSON
    kong.log.info("ASCEND_DECISION: ", cjson.encode(log_entry))
end

return AscendHandler

File: ascend-envoy-authz/internal/authz/handler.go
// Package authz implements the Envoy ext_authz Authorization service.
//
// FAIL SECURE DESIGN:
//   - All errors result in request denial
//   - Only explicitly approved actions proceed
//   - Circuit breaker prevents cascade failures
//
// Author: ASCEND Platform Engineering
package authz

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
	authv3 "github.com/envoyproxy/go-control-plane/envoy/service/auth/v3"
	typev3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
	"google.golang.org/genproto/googleapis/rpc/status"
	"google.golang.org/grpc/codes"

	"github.com/owkai/ascend-envoy-authz/internal/ascend"
	"github.com/owkai/ascend-envoy-authz/internal/config"
	"github.com/owkai/ascend-envoy-authz/internal/logger"
	"github.com/owkai/ascend-envoy-authz/internal/mapper"
)

// Handler implements the Envoy ext_authz Authorization service.
type Handler struct {
	authv3.UnimplementedAuthorizationServer

	client   *ascend.Client
	config   *config.Config
	logger   *logger.Logger

	// Decision cache (approved only)
	cache   map[string]*cacheEntry
	cacheMu sync.RWMutex
}

type cacheEntry struct {
	decision  *ascend.Decision
	expiresAt time.Time
}

// NewHandler creates a new authorization handler.
func NewHandler(cfg *config.Config, client *ascend.Client) *Handler {
	h := &Handler{
		client: client,
		config: cfg,
		logger: logger.Default().WithField("component", "authz_handler"),
		cache:  make(map[string]*cacheEntry),
	}

	// Start cache cleanup goroutine
	go h.cleanupCache()

	h.logger.Info("Authorization handler initialized",
		"fail_open", cfg.FailOpen,
		"cache_ttl", cfg.CacheTTL.String(),
		"agent_id_header", cfg.AgentIDHeader,
	)

	return h
}

// Check implements the Authorization.Check RPC.
// This is the main entry point for Envoy external authorization.
func (h *Handler) Check(ctx context.Context, req *authv3.CheckRequest) (*authv3.CheckResponse, error) {
	startTime := time.Now()

	// Extract HTTP request attributes
	httpReq := req.GetAttributes().GetRequest().GetHttp()
	if httpReq == nil {
		h.logger.Error("Invalid request: missing HTTP attributes")
		return h.denyResponse("Invalid request: missing HTTP attributes", codes.InvalidArgument, nil), nil
	}

	path := httpReq.GetPath()
	method := httpReq.GetMethod()

	// Remove query string for path matching
	if idx := strings.Index(path, "?"); idx != -1 {
		path = path[:idx]
	}

	// Check path exclusions
	if mapper.IsPathExcluded(path, h.config.ExcludedPaths) {
		h.logger.Debug("Path excluded from governance", "path", path)
		return h.allowResponse(nil), nil
	}

	// Extract agent ID from headers
	agentID := h.extractAgentID(httpReq.GetHeaders())
	if agentID == "" {
		if h.config.RequireAgentID {
			h.logger.Warn("Missing agent ID header",
				"header", h.config.AgentIDHeader,
				"path", path,
			)
			return h.denyResponse(
				fmt.Sprintf("Missing required header: %s", h.config.AgentIDHeader),
				codes.Unauthenticated,
				nil,
			), nil
		}
		// Not required, skip governance
		h.logger.Debug("No agent ID, skipping governance", "path", path)
		return h.allowResponse(nil), nil
	}

	// Map request to ASCEND action
	action := mapper.MapCheckRequest(req, agentID, h.config.Environment, h.config.DataSensitivity)

	// Check cache first
	cacheKey := mapper.GenerateCacheKey(agentID, method, path)
	if h.config.CacheTTL > 0 {
		if cached := h.getFromCache(cacheKey); cached != nil {
			h.logger.Debug("Cache hit",
				"agent_id", agentID,
				"cache_key", cacheKey,
			)
			return h.buildResponse(cached), nil
		}
	}

	// Call ASCEND API
	decision, err := h.client.EvaluateAction(ctx, action)
	if err != nil {
		h.logger.Error("ASCEND API error",
			"error", err,
			"agent_id", agentID,
			"path", path,
		)

		// FAIL SECURE: Deny on error unless fail_open is configured
		if h.config.FailOpen {
			h.logger.Warn("Fail open enabled, allowing request despite error")
			return h.allowResponse(map[string]string{
				"x-ascend-status": "error-allowed",
				"x-ascend-error":  err.Error(),
			}), nil
		}

		return h.denyResponse(
			"Governance service unavailable",
			codes.Unavailable,
			map[string]string{"x-ascend-error": "service_unavailable"},
		), nil
	}

	// Cache approved decisions only
	if decision.IsApproved() && h.config.CacheTTL > 0 {
		h.setCache(cacheKey, decision)
	}

	// Log decision
	latencyMs := float64(time.Since(startTime).Microseconds()) / 1000
	if h.config.LogDecisions {
		h.logger.Info("Authorization decision",
			"agent_id", agentID,
			"action_type", action.ActionType,
			"status", decision.Status,
			"risk_score", decision.RiskScore,
			"risk_level", decision.RiskLevel,
			"action_id", decision.ActionID,
			"latency_ms", latencyMs,
			"cached", false,
		)
	}

	return h.buildResponse(decision), nil
}

// extractAgentID extracts the agent ID from request headers.
func (h *Handler) extractAgentID(headers map[string]string) string {
	// Try exact match first
	if agentID, ok := headers[h.config.AgentIDHeader]; ok {
		return agentID
	}

	// Try case-insensitive match
	headerLower := strings.ToLower(h.config.AgentIDHeader)
	for k, v := range headers {
		if strings.ToLower(k) == headerLower {
			return v
		}
	}

	// Use default if configured
	return h.config.DefaultAgentID
}

// buildResponse converts an ASCEND decision to a CheckResponse.
func (h *Handler) buildResponse(decision *ascend.Decision) *authv3.CheckResponse {
	switch decision.Status {
	case ascend.StatusApproved:
		return h.allowResponse(map[string]string{
			"x-ascend-status":     "approved",
			"x-ascend-action-id":  fmt.Sprintf("%d", decision.ActionID),
			"x-ascend-risk-score": fmt.Sprintf("%.1f", decision.RiskScore),
			"x-ascend-risk-level": decision.RiskLevel,
		})

	case ascend.StatusPending:
		if h.config.BlockOnPending {
			return h.denyResponse(
				"Action requires human approval",
				codes.PermissionDenied,
				map[string]string{
					"x-ascend-status":     "pending_approval",
					"x-ascend-action-id":  fmt.Sprintf("%d", decision.ActionID),
					"x-ascend-risk-score": fmt.Sprintf("%.1f", decision.RiskScore),
				},
			)
		}
		// Allow but flag as pending
		return h.allowResponse(map[string]string{
			"x-ascend-status":    "pending_approval",
			"x-ascend-action-id": fmt.Sprintf("%d", decision.ActionID),
		})

	default: // denied or unknown
		message := "Action denied by governance policy"
		if decision.DenialReason != "" {
			message = decision.DenialReason
		}
		return h.denyResponse(
			message,
			codes.PermissionDenied,
			map[string]string{
				"x-ascend-status":     "denied",
				"x-ascend-action-id":  fmt.Sprintf("%d", decision.ActionID),
				"x-ascend-risk-score": fmt.Sprintf("%.1f", decision.RiskScore),
			},
		)
	}
}

// allowResponse creates an OkHttpResponse (allow the request).
func (h *Handler) allowResponse(headers map[string]string) *authv3.CheckResponse {
	var headersList []*corev3.HeaderValueOption
	for k, v := range headers {
		headersList = append(headersList, &corev3.HeaderValueOption{
			Header: &corev3.HeaderValue{Key: k, Value: v},
		})
	}

	return &authv3.CheckResponse{
		Status: &status.Status{Code: int32(codes.OK)},
		HttpResponse: &authv3.CheckResponse_OkResponse{
			OkResponse: &authv3.OkHttpResponse{
				Headers: headersList,
			},
		},
	}
}

// denyResponse creates a DeniedHttpResponse (block the request).
func (h *Handler) denyResponse(message string, code codes.Code, headers map[string]string) *authv3.CheckResponse {
	var headersList []*corev3.HeaderValueOption
	for k, v := range headers {
		headersList = append(headersList, &corev3.HeaderValueOption{
			Header: &corev3.HeaderValue{Key: k, Value: v},
		})
	}

	// Determine HTTP status code
	httpStatus := typev3.StatusCode_Forbidden
	if code == codes.Unavailable {
		httpStatus = typev3.StatusCode_ServiceUnavailable
	} else if code == codes.Unauthenticated {
		httpStatus = typev3.StatusCode_Unauthorized
	}

	return &authv3.CheckResponse{
		Status: &status.Status{
			Code:    int32(code),
			Message: message,
		},
		HttpResponse: &authv3.CheckResponse_DeniedResponse{
			DeniedResponse: &authv3.DeniedHttpResponse{
				Status: &typev3.HttpStatus{
					Code: httpStatus,
				},
				Headers: headersList,
				Body:    fmt.Sprintf(`{"error":"%s","message":"%s"}`, code.String(), message),
			},
		},
	}
}

// Cache methods

func (h *Handler) getFromCache(key string) *ascend.Decision {
	h.cacheMu.RLock()
	defer h.cacheMu.RUnlock()

	entry, ok := h.cache[key]
	if !ok || time.Now().After(entry.expiresAt) {
		return nil
	}
	return entry.decision
}

func (h *Handler) setCache(key string, decision *ascend.Decision) {
	h.cacheMu.Lock()
	defer h.cacheMu.Unlock()

	h.cache[key] = &cacheEntry{
		decision:  decision,
		expiresAt: time.Now().Add(h.config.CacheTTL),
	}
}

func (h *Handler) cleanupCache() {
	ticker := time.NewTicker(time.Minute)
	defer ticker.Stop()

	for range ticker.C {
		h.cacheMu.Lock()
		now := time.Now()
		for k, v := range h.cache {
			if now.After(v.expiresAt) {
				delete(h.cache, k)
			}
		}
		h.cacheMu.Unlock()
	}
}
